proof/refine/ARM/Reply_R.thy

```diff
@@ -634,4 +586,426 @@ lemma replyRemoveTCB_invs':
               simp: cteCaps_of_def)
   done
 
+context begin interpretation Arch .
+
+lemma pspace_relation_reply_update_conc_only:
+  "\<lbrakk> pspace_relation ps ps'; ps x = Some (Structures_A.Reply reply); ps' x = Some (KOReply reply');
+     reply_relation reply new\<rbrakk>
+   \<Longrightarrow> pspace_relation ps (ps'(x \<mapsto> (KOReply new)))"
+  apply (simp add: pspace_relation_def pspace_dom_update dom_fun_upd2
+              del: dom_fun_upd)
+  apply (erule conjE)
+  apply (rule ballI, drule(1) bspec)
+  apply (clarsimp simp: split_def)
+  apply (drule bspec, simp)
+  apply (clarsimp simp: obj_relation_cuts_def2 cte_relation_def
+                        pte_relation_def pde_relation_def
+                 split: Structures_A.kernel_object.split_asm arch_kernel_obj.split_asm if_split_asm)
+  done
+
+end
+
+lemma replyPrevs_of_replyNext_update:
+  "ko_at' reply' rp s' \<Longrightarrow>
+      replyPrevs_of (s'\<lparr>ksPSpace := ksPSpace s'(rp \<mapsto>
+                            KOReply (reply' \<lparr> replyNext := v \<rparr>))\<rparr>) = replyPrevs_of s'"
+  apply (clarsimp simp: obj_at'_def projectKOs isNext_def
+                 split: option.split_asm reply_next.split_asm)
+  by (fastforce simp: projectKO_opt_reply opt_map_def)
+
+lemma scs_of'_reply_update:
+  "reply_at' rp s' \<Longrightarrow>
+      scs_of' (s'\<lparr>ksPSpace := ksPSpace s'(rp \<mapsto> KOReply reply)\<rparr>) = scs_of' s'"
```

Comment:
It's a bit more complex than that, because we do want non-negated `obj_at'` at `rp`. But it is doable, so why not. 
I've replaced it with:
```
lemma scs_of'_non_sc_update:
  "obj_at' (λ_::'a::pspace_storable. True) rp s' ⟹ ¬ sc_at' rp s' ⟹
   koTypeOf (injectKO v) ≠ SchedContextT ⟹
      scs_of' (s'⦇ksPSpace := ksPSpace s'(rp ↦ injectKO v)⦈) = scs_of' s'"
```